pkg/storegateway/indexheader/encoding/factory.go

```diff
@@ -125,18 +163,104 @@ func (df *DecbufFactory) NewRawDecbuf() Decbuf {
-func (df *DecbufFactory) CloseWithErrCapture(err *error, d Decbuf, format string, a ...interface{}) {
-	merr := multierror.MultiError{}
+// filePool maintains a pool of file handles up to a maximum number, creating
+// new handles and closing them when required. get and put operations on this
+// pool never block. If there are no available file handles, one will be created
+// on get. If the pool is full, the file handle is closed on put.
+type filePool struct {
+	path    string
+	logger  log.Logger
+	handles chan *os.File
+	mtx     sync.RWMutex
+	stopped bool
+
+	opens        prometheus.Counter
+	pooledOpens  prometheus.Counter
+	closes       prometheus.Counter
+	pooledCloses prometheus.Counter
+}
 
-	merr.Add(*err)
-	merr.Add(errors.Wrapf(df.Close(d), format, a...))
+// newFilePool creates a new file pool for path with cap capacity. If cap is 0,
+// get always opens new file handles and put always closes them immediately.
+func newFilePool(path string, cap uint, logger log.Logger, opens prometheus.Counter, pooledOpens prometheus.Counter, closes prometheus.Counter, pooledCloses prometheus.Counter) *filePool {
+	return &filePool{
+		path:   path,
+		logger: logger,
+		// We don't care if cap is 0 which means the channel will be unbuffered. Because
+		// we have default cases for reads and writes to the channel, we will always open
+		// new files and close file handles immediately if the channel is unbuffered.
+		handles: make(chan *os.File, cap),
+
+		opens:        opens,
+		pooledOpens:  pooledOpens,
+		closes:       closes,
+		pooledCloses: pooledCloses,
+	}
+}
+
+// get returns a pooled file handle if available or opens a new one if there
+// are no pooled handles available. If this pool has been stopped, an error
+// is returned.
+func (p *filePool) get() (*os.File, error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	if p.stopped {
+		return nil, ErrPoolStopped
+	}
 
-	*err = merr.Err()
+	select {
+	case f := <-p.handles:
+		p.pooledOpens.Inc()
+		return f, nil
+	default:
+		p.opens.Inc()
+		return os.Open(p.path)
+	}
+}
+
+// put returns a file handle to the pool if there is space available or closes
+// the file handle if there is not. If this pool has been stopped, the file handle
+// is closed immediately.
+func (p *filePool) put(f *os.File) error {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	if p.stopped {
+		return f.Close()
+	}
+
+	select {
+	case p.handles <- f:
+		p.pooledCloses.Inc()
+		return nil
+	default:
+		p.closes.Inc()
+		return f.Close()
+	}
+}
+
+// stop closes all pooled file handles. After this method is called, subsequent
+// get calls will return an error and put calls will immediately close the file
+// handle.
+func (p *filePool) stop() {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+	p.stopped = true
+
+	for {
+		select {
+		case f := <-p.handles:
+			if err := f.Close(); err != nil {
+				level.Warn(p.logger).Log("msg", "closing index-header files during pool stop", "path", p.path, "err", err)
```

Comment:
[nit] Given we'll keep trying to close other file handles, this seems clearer to me

```suggestion
				level.Warn(p.logger).Log("msg", "closing index-header file during pool stop", "path", p.path, "err", err)
```